[CORE-10604] schema_registry/authz: Extend ACL resources

BenPope opened the pull request:
Introduce the concept of a `subsystem` into ACLs, so that `kafka` resources and `schema_registry` resources can operate independently.

ACLs for `kafka` resources are managed by the Kafka API, and operate on Kafka resources.
ACLs for `schema_registry` resources shall be managed through a separate API.

It is important that the two subsystems do not interact with each other.

Replaces #26388

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

BenPope committed:
schema_registry/authz: Introduce global_resource

Signed-off-by: Ben Pope <ben@redpanda.com>

pgellert commented on src/v/kafka/server/handlers/details/security.h:
```diff
@@ -419,6 +439,10 @@ const std::vector<security::acl_operation>& get_allowed_operations() {
         return topic_resource_ops;
     case security::resource_type::transactional_id:
         return transactional_id_resource_ops;
+    case security::resource_type::sr_subject:
+        return sr_subject_resource_ops;
+    case security::resource_type::sr_global:
+        return sr_global_resource_ops;
```
As far as I can tell this feeds SR ACLs into the metadata/describe_cluster response's authorized ops. Isn't that something we would want to avoid?

(+ I think we would want to avoid any SR ACL-related logic in the kafka module, so if this is necessary, maybe we should move it into a common place like security.)

BenPope replied:
> As far as I can tell this feeds SR ACLs into the metadata/describe_cluster

It requests `get_allowed_operations<security::acl_cluster_name>`, which is mapped to `security::resource_type::cluster` (Line 434), so SR resources are not returned via the Kafka API.

pgellert commented on src/v/security/acl.cc:
```diff
@@ -498,6 +515,19 @@ bool resource_pattern_filter::matches(const resource_pattern& pattern) const {
+            return false;
+        }
+        break;
+    case resource_subsystem::schema_registry:
+        if (pattern.resource() < resource_type::sr_subject) {
+            return false;
+        }
+        break;
+    }
+
```
Should `resource_type::cluster` be included in both the kafka subsystem as well as the SR subsystem? If I understand the RFC correctly, `describe/alter cluster` permissions would apply to both the kafka and SR subsystems.

pgellert replied:
Do we have product alignment on whether or not to share the cluster-level permissions across SR and kafka?

BenPope replied:
> Should `resource_type::cluster` be included in both the kafka subsystem as well as the SR subsystem?

No:
* `resource_type::sr_global` is used to lock down SR resources via the SR REST API proper.
* `resource_type::cluster` is used to lock the `security/acl` endpoints on the SR `security/acl` REST API.

pgellert replied:
Exactly, given that, I would expect that if I filter for the SR subsystem I would get back `resource_type::cluster` as well, since `resource_type::cluster` is a SR-impacting ACL.

I imagine `GET /security/acls` will use `resource_pattern_filter::any(resource_subsystem::schema_registry)` to find the ACLs to return. Won't we want to return `resource_type::cluster` on that endpoint as well? To me it seems like we should, otherwise users can't see through the SR API who has cluster-level priviledges. Am I missing something here?

BenPope replied:
Oh, I see.

I was thinking that `security/acl` would query `resource_subsystem::kafka` for permissions over `resource::type::cluster` to access `security/acl` endpoints (unless it's a `superuser`).

That way, it's possible for a `superuser` or cluster admin to grant permissions for SR ACL management.

I don't think it's necessary to `describe_config` or `alter_config` permissions over `resource_type::cluster` using the SR admin API. We don't want to duplicate administration of permissions over any of the `resource_subsystem::kafka` resources; that should be done over the Kafka API.

So `security/acl` will require `resource_subsystem::kafka`, but return `resource_subsystem::schema_registry` ACLs.

michael-redpanda replied:
Good eye @pgellert....

Yeah I can see this being somewhat confusing for a customer - you have to use the Kafka API (or Admin API for superusers) to grant permissions to modify SR ACLs...

At least if they're using `rpk` then it's still the same tool (they don't have to know it's hitting two different endpoints).

Let's talk at stand up tomorrow.  I think we have a few options:

1. Leave it as designed
2. Permit viewing cluster resources over this API
  a. Also permit adding ACLs to cluster resources over this API
3. Have a different resource (maybe the global one?) be the ACL that controls this endpoint

pgellert commented on src/v/kafka/server/handlers/details/security.h:
```diff
@@ -270,6 +270,10 @@ inline int8_t to_kafka_resource_type(security::resource_type type) {
         return 5;
+    case security::resource_type::sr_subject:
+    case security::resource_type::sr_global:
+        vassert(
+          false, "Schema Registry resources are not supported in kafka ACLs");
```
Do we need any new validation of kafka request inputs to prevent ever hitting this case?

dotnwat replied:
great question

BenPope replied:
> Do we need any new validation of kafka request inputs to prevent ever hitting this case?

The delete/describe acl request requires a resource type with kafka enum values, which are already validated in `kafka::details::to_resource_type`, which throws an `acl_conversion_error`.

dotnwat commented on src/v/security/acl.cc:
```diff
@@ -558,6 +588,11 @@ void read_nested(
         filter._pattern = security::resource_pattern_filter::pattern_match{};
         break;
     }
+    if (in.bytes_left() > 0) {
```
should we be bumping the serde version instead?

BenPope replied:
> should we be bumping the serde version instead?

I don't know how to do that, the version is not available in `read_nested`.

BenPope replied:
There does seem to be an issue with serde, though.

dotnwat replied:
>I don't know how to do that, the version is not available in read_nested.

i think that is by design: read_nested is low-level, DIY serialization so you have to read the header containing the versions manually. but upon closer inspection, it looks like when serialization was added resource_pattern_filter that it may have been added in a way that treated resource_pattern_filter as a primitive (e.g. like std::vector or whatever) which contains no version information.

i'm pretty sure the version information isn't on the wire, but we need to double check that. if it is indeed missing, then we'll have to get creative (e.g. by hijacking an unused byte some where -- not pretty).

BenPope replied:
I've bumped the version of the containing type and passed that through.

For `reflection::adl<>`, the new field is not added.

dotnwat reviewed (changes requested):
need to resolve the serde issue with acl type

BenPope committed:
security/acl: Encapsulate serde

The serialisation is only needed in the implementation file.

This simplifies the next commit.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
security/acl: Introduce resource_patter_filter::resource_subsystem

ACLs currently assume everything is a kafka resource.

Introduce a mechanism to distinguish kafka resources and
schema registry resources; they operate independently.

Since `resource_pattern_filter` does not have a header, the version
from the containing type, `acl_binding_filter` is bumped, and passed
through to `resource_pattern_filter::nested_read`, which is no longer
compatible with serde.

`reflection::adl` always assumes `resource_subsystem::kafka`.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
security: Add support for SR resources.

Add support for two new resources:
* ppsr::subject
* ppsr::global_resource

Signed-off-by: Ben Pope <ben@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--fe858ae3-8943-4bd7-94e0-5155b5a330f2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67170">build#67170</a></summary>


| test_class                   | test_method                                       | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:--------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigAliasTest       | test_aliasing                                     | {"prop_set": ["data_transforms_per_core_memory_reservation", "wasm_per_core_memory_reservation", [23, 3], [27000000, 37000000, 47000000], true]} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67170#01975f0c-ea7b-4cee-b4ec-22f869390027 | FAIL          | 0/1      |                                                                                                                                                                         |
| NodeWiseRecoveryTest         | test_node_wise_recovery                           | {"dead_node_count": 1}                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67170#01975f0a-e0e9-438a-b1fc-963ee2c0d7e8 | FLAKY         | 16/19    |                                                                                                                                                                         |
| PartitionReassignmentsTest   | test_add_partitions_with_inprogress_reassignments |                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67170#01975f0a-e0ea-49f3-88b8-ffa6774b19d3 | FLAKY         | 16/21    | upstream reliability is '94.47852760736197'. current run reliability is '76.19047619047619'. drift is 18.28805 and the allowed drift is set to 50. The test should PASS |
| UsageTestCloudStorageMetrics | test_usage_manager_cloud_storage                  |                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67170#01975f0a-e0ea-49f3-88b8-ffa6774b19d3 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
</details>
<!--/body-->

dotnwat commented on src/v/security/acl.h:
```diff
@@ -403,11 +408,17 @@ class acl_binding
 /*
  * A filter for matching resources.
+ *
+ * version 1 adds subsystem.
+ *
+ * Note: This type does not have a serde::header, the version must be provided
+ * by the caller of nested_read, which has the wrong signature for serde.
+ *
  */
 class resource_pattern_filter
   : public serde::envelope<
       resource_pattern_filter,
-      serde::version<0>,
+      serde::version<1>,
```
there is no header, and never will be. so i think it would make sense to remove the envelope inheritence here? otherwise, it seems rather confusing. resource_pattern_filter suffers the fate of requiring custom serialization now.

dotnwat commented on src/v/security/acl.cc:
```diff
@@ -529,7 +564,8 @@ bool resource_pattern_filter::matches(const resource_pattern& pattern) const {
 void read_nested(
   iobuf_parser& in,
   resource_pattern_filter& filter,
-  const size_t bytes_left_limit) {
+  const size_t bytes_left_limit,
+  serde::version_t version) {
```
I think that life is going to be much nicer to us if we can contain the serde complexity within `resource_pattern_filter` rather than relying on its including type to provide versioning information (it's not a bad trick, though--i just think we should make it our last option).

What we want is to bootstrap some kind of version into the serialization protocol for this type. To do that we need a reliable way to identify instances of the unversioned struct, and implicity treat that as the first version.

Probably the most straight forward signal to extract is that `bool` on the wire (the prefix on std::optional fields) is going to be 0 or 1. However, serde will decode any non-zero value as true. We may also be able to write a sential value onto the wire that new code can look for (will old code ignore it properly?).

Pretty sure @mmaslankaprv has deployed these tricks before.

Did you consider these or any other tricks?

michael-redpanda replied:
I want to ask a possibly dumb question - how do we _know_ that the envelope for `resource_pattern_filter` is _not_ being written?  What pattern in the code is suggesting that?

`acl_binding_filter` uses `serde_fields` wouldn't that cause `serde` to treat `resource_pattern_filter` as an envelope?

michael-redpanda replied:
Also do we need a feature flag for this?

michael-redpanda replied:
n/m I know now

dotnwat replied:
> n/m I know now

what did you find out?

michael-redpanda commented on src/v/security/acl.h:
```diff
@@ -50,10 +48,13 @@ enum class resource_type : int8_t {
     group = 1,
     cluster = 2,
     transactional_id = 3,
+    sr_subject = 4,
+    sr_global = 5,
```
Can we make these like either starting at 10 or 100?  just incase Kafka adds new resource types at 4 & 5

BenPope replied:
> Can we make these like either starting at 10 or 100?  just incase Kafka adds new resource types at 4 & 5

We can, but the types and values are not directly correlated; they have explicit mappings in each direction (they just happen to have the same values).

michael-redpanda replied:
ah ok so these are not the Kafka protocol, got it

pgellert commented:
Moving the implementation of this over to a new PR (since I took over finishing this and couldn't push to Ben's fork): https://github.com/redpanda-data/redpanda/pull/26466
